plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/Inline.hs

```diff
@@ -326,14 +326,50 @@ sizeIsAcceptable = \case
   Delay _ t  -> sizeIsAcceptable t
 
 isPure :: Term name uni fun a -> Bool
-isPure = go
+isPure = go True
     where
-        go = \case
-            Var {}      -> True
+        -- See Note [delayAndVarIsPure]
+        go delayAndVarIsPure = \case
+            Var {}                        -> delayAndVarIsPure
             -- These are syntactically values that won't reduce further
-            LamAbs {}   -> True
-            Constant {} -> True
-            Delay {}    -> True
-
+            LamAbs {}                     -> True
+            Constant {}                   -> True
+            Delay _ body                  -> delayAndVarIsPure || go delayAndVarIsPure body
+            -- This case is not needed in PIR's `isPure`, because PIR's beta-reduction pass
+            -- turns terms like this into `Let` bindings.
+            Apply _ (LamAbs _ _ body) arg -> go True arg && go delayAndVarIsPure body
```

Comment:
This case is actually quite subtle, I had to stare at it for a while to work out why it's okay. Worth a comment, and does one of the tests witness that it does the right thing?